src/dsa/rpo_falcon512/signature.rs

```diff
@@ -39,251 +42,334 @@ use super::{
-    /// of the nonce and interpreting them as field elements.
-    pub fn nonce(&self) -> NonceElements {
-        // we assume that the signature was constructed with a valid signature, and thus
-        // expect() is OK here.
-        let nonce = self.sig[1..41].try_into().expect("invalid signature");
-        decode_nonce(nonce)
-    }
+    // PUBLIC ACCESSORS
+    // --------------------------------------------------------------------------------------------
 
     // Returns the polynomial representation of the signature in Z_p[x]/(phi).
-    pub fn sig_poly(&self) -> Polynomial {
-        *self.sig_polynomial.get_or_init(|| {
-            // we assume that the signature was constructed with a valid signature, and thus
-            // expect() is OK here.
-            Polynomial::from_signature(&self.sig).expect("invalid signature")
-        })
+    pub fn sig_poly(&self) -> (&Polynomial<FalconFelt>, &Polynomial<FalconFelt>) {
+        (&self.s1, &self.s2)
     }
 
-    // HASH-TO-POINT
-    // --------------------------------------------------------------------------------------------
-
-    /// Returns a polynomial in Z_p\[x\]/(phi) representing the hash of the provided message.
-    pub fn hash_to_point(&self, message: Word) -> Polynomial {
-        hash_to_point(message, &self.nonce())
+    /// Returns the nonce component of the signature.
+    pub fn nonce(&self) -> &Nonce {
+        &self.nonce
     }
 
     // SIGNATURE VERIFICATION
     // --------------------------------------------------------------------------------------------
+
     /// Returns true if this signature is a valid signature for the specified message generated
-    /// against key pair matching the specified public key commitment.
+    /// against the secret key matching the specified public key commitment.
     pub fn verify(&self, message: Word, pubkey_com: Word) -> bool {
-        // Make sure the expanded public key matches the provided public key commitment
-        let h = self.pub_key_poly();
-        let h_digest: Word = Rpo256::hash_elements(&h.to_elements()).into();
-        if h_digest != pubkey_com {
-            return false;
-        }
+        let c = hash_to_point_rpo256(message, &self.nonce);
+        let s1 = self.s1.clone();
+        let s2 = self.s2.clone();
+        let s1_fft = s1.fft();
+        let s2_fft = s2.fft();
+        let c_fft = c.fft();
 
-        // Make sure the signature is valid
-        let s2 = self.sig_poly();
-        let c = self.hash_to_point(message);
+        // h = s2^(-1) * (c - s1)
+        let h_fft = (c_fft - s1_fft).hadamard_div(&s2_fft);
+        let h = h_fft.ifft();
 
-        let s1 = c - s2 * h;
+        let h_felt: Polynomial<Felt> = h.clone().into();
+        let h_digest: Word = Rpo256::hash_elements(&h_felt.coefficients).into();
 
-        let sq_norm = s1.sq_norm() + s2.sq_norm();
-        sq_norm <= SIG_L2_BOUND
+        h_digest == pubkey_com && verify_helper(c, s2, h.into())
```

Comment:
Let's add some simple comments here - e.g., "recover public key from the signature", "compute hash of the public key" etc.